File: Kalorie/Kalorie/Core/UseCases/FetchFoodByBarcodeExternallyUseCase.swift
//
//  FetchFoodByBarcodeExternallyUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 24.07.2026.
//

import Foundation

enum FetchFoodByBarcodeExternallyError: Error {
    case invalidURL
    case serverError(statusCode: Int)
}

protocol FetchFoodByBarcodeExternallyUseCaseProtocol {
    func callAsFunction(barcode: String) async throws -> FoodItemDomain?
}

struct FetchFoodByBarcodeExternallyUseCase: FetchFoodByBarcodeExternallyUseCaseProtocol {

    // MARK: - Properties

    private let session: URLSession
    private let retryDelay: Duration

    // MARK: - Init

    init(session: URLSession = .shared, retryDelay: Duration = Constants.OpenFoodFacts.retryDelay) {
        self.session = session
        self.retryDelay = retryDelay
    }

    // MARK: - Functions

    func callAsFunction(barcode: String) async throws -> FoodItemDomain? {
        guard !barcode.isEmpty else { return nil }
        guard let encodedBarcode = barcode.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)
        else { throw FetchFoodByBarcodeExternallyError.invalidURL }
        var components = URLComponents()
        components.scheme = "https"
        components.host = Constants.OpenFoodFacts.host
        components.percentEncodedPath = "/api/v2/product/\(encodedBarcode)"
        components.queryItems = [
            URLQueryItem(name: "fields", value: "code,product_name,product_name_cs,product_name_en,nutriments")
        ]
        guard let url = components.url else { throw FetchFoodByBarcodeExternallyError.invalidURL }
        // URLSession is used directly — OpenFoodFacts is plain HTTP, not Firestore, so FirestoreDataProviderProtocol doesn't apply.
        // The trade-off: fakes can only stub the return value, not assert which URL was called.
        var request = URLRequest(url: url, timeoutInterval: Constants.OpenFoodFacts.requestTimeout)
        request.setValue(Constants.OpenFoodFacts.userAgent, forHTTPHeaderField: "User-Agent")
        let (data, statusCode) = try await OpenFoodFactsTransientRequest.data(for: request, session: session, retryDelay: retryDelay)
        guard (200...299).contains(statusCode) else {
            throw FetchFoodByBarcodeExternallyError.serverError(statusCode: statusCode)
        }
        let decoded = try JSONDecoder().decode(OpenFoodFactsBarcodeResponseDTO.self, from: data)
        guard decoded.status == 1, let product = decoded.product else { return nil }
        return product.asDomain()
    }

}

#if DEBUG
struct FetchFoodByBarcodeExternallyUseCaseFake: FetchFoodByBarcodeExternallyUseCaseProtocol {

    // MARK: - Properties

    var stubbedItem: FoodItemDomain?
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(barcode: String) async throws -> FoodItemDomain? {
        if shouldThrow { throw FetchFoodByBarcodeExternallyError.invalidURL }
        return stubbedItem
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchFoodByBarcodeExternallyUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.networking.HttpRequest
import antoni.kalorie.core.networking.HttpSessionProtocol
import antoni.kalorie.core.networking.HttpURLConnectionSession
import antoni.kalorie.core.networking.OpenFoodFactsBarcodeResponseDTO
import antoni.kalorie.core.networking.OpenFoodFactsTransientRequest
import antoni.kalorie.core.utils.Constants
import java.net.URLEncoder
import kotlin.time.Duration
import kotlinx.serialization.json.Json

sealed class FetchFoodByBarcodeExternallyError : Exception() {
    data object InvalidURL : FetchFoodByBarcodeExternallyError()
    data class ServerError(val statusCode: Int) : FetchFoodByBarcodeExternallyError()
}

interface FetchFoodByBarcodeExternallyUseCaseProtocol {
    suspend operator fun invoke(barcode: String): FoodItemDomain?
}

class FetchFoodByBarcodeExternallyUseCase(
    private val session: HttpSessionProtocol = HttpURLConnectionSession(),
    private val retryDelay: Duration = Constants.OpenFoodFacts.RETRY_DELAY,
) : FetchFoodByBarcodeExternallyUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(barcode: String): FoodItemDomain? {
        if (barcode.isEmpty()) return null
        val encodedBarcode = URLEncoder.encode(barcode, Charsets.UTF_8).replace("+", "%20")
        val request = HttpRequest(
            url = "https://${Constants.OpenFoodFacts.HOST}/api/v2/product/$encodedBarcode?fields=$FIELDS",
            userAgent = Constants.OpenFoodFacts.USER_AGENT,
            timeoutMillis = Constants.OpenFoodFacts.REQUEST_TIMEOUT_MILLIS,
        )
        val response = OpenFoodFactsTransientRequest.data(request, session, retryDelay)
        if (response.statusCode !in 200..299) throw FetchFoodByBarcodeExternallyError.ServerError(response.statusCode)
        val decoded = json.decodeFromString<OpenFoodFactsBarcodeResponseDTO>(response.body)
        val product = decoded.product
        if (decoded.status != 1 || product == null) return null
        return product.asDomain()
    }

    private companion object {
        const val FIELDS = "code,product_name,product_name_cs,product_name_en,nutriments"
        val json = Json { ignoreUnknownKeys = true }
    }
}
